### PR TITLE
[SPARK-19918][SQL] Use TextFileFormat in implementation of TextInputJsonDataSource

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR proposes to use text datasource when Json schema inference.

This basically proposes the similar approach in https://github.com/apache/spark/pull/15813 If we use Dataset for initial loading when inferring the schema, there are advantages. Please refer SPARK-18362

It seems JSON one was supposed to be fixed together but taken out according to https://github.com/apache/spark/pull/15813

> A similar problem also affects the JSON file format and this patch originally fixed that as well, but I've decided to split that change into a separate patch so as not to conflict with changes in another JSON PR.

Also, this seems affecting some functionalities because it does not use `FileScanRDD`. This problem is described in SPARK-19885 (but it was CSV's case).

## How was this patch tested?

Existing tests should cover this and manual test by `spark.read.json(path)` and check the UI.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
(let me wait for the tests before cc'ing someone)

**Comment by HyukjinKwon** [comment]
cc @cloud-fan, @joshrosen and @NathanHowell  could you take a look and see if it makes sense when you have some time?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonDataSource.scala [hunk 1]
```diff
@@ -23,24 +23,25 @@ import com.fasterxml.jackson.core.{JsonFactory, JsonParser}
...
+import org.apache.hadoop.io.Text
 import org.apache.hadoop.mapreduce.Job
-import org.apache.hadoop.mapreduce.lib.input.{FileInputFormat, TextInputFormat}
+import org.apache.hadoop.mapreduce.lib.input.FileInputFormat
 
 import org.apache.spark.TaskContext
 import org.apache.spark.input.{PortableDataStream, StreamInputFormat}
 import org.apache.spark.rdd.{BinaryFileRDD, RDD}
-import org.apache.spark.sql.{AnalysisException, SparkSession}
+import org.apache.spark.sql.{AnalysisException, Encoders, SparkSession}
```
Comment (NathanHowell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonDataSource.scala]: Is the `Encoders` import still necessary?

**Comment by NathanHowell** [comment]
Would there be any additional benefit of replacing more (or all?) of the uses of `RDD` with the equivalent `Dataset` operations?

**Comment by HyukjinKwon** [comment]
Let me take a look for that. I think we can replace `sample` at least.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonDataSource.scala [hunk 2]
```diff
@@ -123,24 +124,31 @@ object JsonDataSource {
...
+        val df = sparkSession.baseRelationToDataFrame(
+          DataSource.apply(
+            sparkSession,
+            paths = paths,
+            className = classOf[TextFileFormat].getName
+          ).resolveRelation(checkFilesExist = false))
+          .select("value")
+
+        // Avoid unnecessary conversion `UTF8String` into string.
+        val physicalPlan = df.queryExecution.executedPlan
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonDataSource.scala]: `df.queryExecution.toRDD`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonDataSource.scala [hunk 3]
```diff
@@ -123,24 +124,31 @@ object JsonDataSource {
...
+object TextInputJsonDataSource extends JsonDataSource[UTF8String] {
   override val isSplitable: Boolean = {
     // splittable if the underlying source is
     true
   }
 
   override protected def createBaseRdd(
       sparkSession: SparkSession,
-      inputPaths: Seq[FileStatus]): RDD[Text] = {
+      inputPaths: Seq[FileStatus]): RDD[UTF8String] = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonDataSource.scala]: can we make it return `Dataset` like CSV?

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonDataSource.scala]: Let me try.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonUtils.scala [hunk 4]
```diff
@@ -0,0 +1,50 @@
+/*
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonUtils.scala]: This might be too much. I am willing to revert this back if anyone feels it is a bit odd.

I just made this just for matching it to `CSVUtils` that contains variants for logically same preprocessing performed on different data (e.g. `Iterator`, `RDD`, `Dataset`).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonDataSource.scala [hunk 5]
```diff
@@ -17,32 +17,30 @@
...
 import org.apache.spark.sql.types.StructType
 import org.apache.spark.unsafe.types.UTF8String
 import org.apache.spark.util.Utils
 
 /**
  * Common functions for parsing JSON files
- * @tparam T A datatype containing the unparsed JSON, such as [[Text]] or [[String]]
  */
-abstract class JsonDataSource[T] extends Serializable {
+abstract class JsonDataSource extends Serializable {
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonDataSource.scala]: The changes in this file basically resembles `CSVDataSource`.  (Note that this is almost identical if https://github.com/apache/spark/pull/17256 is merged).

**Comment by HyukjinKwon** [comment]
It seems all builds were gone failed unexpectedly. Let me restart.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonInferSchema.scala [hunk 6]
```diff
@@ -40,18 +40,11 @@ private[sql] object JsonInferSchema {
...
       createParser: (JsonFactory, T) => JsonParser): StructType = {
-    require(configOptions.samplingRatio > 0,
-      s"samplingRatio (${configOptions.samplingRatio}) should be greater than 0")
     val shouldHandleCorruptRecord = configOptions.permissive
     val columnNameOfCorruptRecord = configOptions.columnNameOfCorruptRecord
-    val schemaData = if (configOptions.samplingRatio > 0.99) {
-      json
-    } else {
-      json.sample(withReplacement = false, configOptions.samplingRatio, 1)
-    }
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonInferSchema.scala]: why move the sample logic out?

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonInferSchema.scala]: Because `JsonInferSchema.infer` takes an `RDD[T]` which is the actual source to parse JSON strings. In case of whole text, it is `RDD[PortableDataStream]` whereas for normal one, it is `RDD[UTF8String]`.

Thing is, it seems there is an advantage of doing the sample operation on `Dataset[String]` (not on `RDD`). So, the sample had to be applied onto `Dataset[String]` before converting it into `RDD[UTF8String]`.

In a simple view:

- `TextInputJsonDataSource`:

  ```scala
  val json: Dataset[String] = ...
  val sampled: Dataset[String] = JsonUtils.sample(...)
  val rdd: RDD[UTF8String] = ...
  JsonInferSchema.infer(rdd)
  ```

- `WholeFileJsonDataSource`:

  ```scala
  val json: RDD[PortableDataStream] = ...
  val sampled: RDD[PortableDataStream] = JsonUtils.sample(...)
  JsonInferSchema.infer(rdd)
  ```

I could not find a good way to generalize `JsonInferSchema.infer` to take both `Dataset` and `RDD` as the source so that keep the logic within here with small and clean changes.

If this question is about why it use `Dataset.sample` instead of `RDD.sample`, it is suggested in https://github.com/apache/spark/pull/17255#issuecomment-285960658.

Up to my knowledge, both use the sample sampler `BernoulliCellSampler` as replacements are disabled but for `Dataset` one, it generates the codes. So, I thought there might be a bit of benefits.

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonInferSchema.scala]: Strictly, maybe this is not directly related with the JIRA. I am willing to revert this change back or please let me know if you have a better idea.

Comment (NathanHowell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonInferSchema.scala]: Yah, perhaps the RDD->Dataset changes should be done under a separate issue. I think it can be done across the board (removing most/all RDD references) but I'm not sure what other implications it would have.

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonInferSchema.scala]: (My worry is though.. it might be a hard work to check each really is better on `Dataset`..) @cloud-fan, if you are not sure of here, let me revert. Please let me know.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JsonInferSchema.scala]: I think it's fine

**Comment by cloud-fan** [comment]
thanks, merging to master!
